packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts

```diff
@@ -0,0 +1,183 @@
+const PEER_ID = 'PEER_ID'
+const QUERIES_PER_SECOND = 5
+const MAX_QUEUED_QUERIES = 10
+
+describe('pubsub with queries rate limited', () => {
+  let ipfs
+  let pubsub
+
+  beforeEach(() => {
+    ipfs = {
+      pubsub: {
+        subscribe: jest.fn(),
+        unsubscribe: jest.fn(),
+        ls: jest.fn(() => []),
+        publish: jest.fn(),
+      },
+      id: jest.fn(async () => ({ id: PEER_ID })),
+    }
+    const raw_pubsub = new Pubsub(ipfs, TOPIC, 3000, pubsubLogger, diagnosticsLogger)
+    pubsub = new PubsubRateLimit(
+      raw_pubsub,
+      new LoggerProvider().getDiagnosticsLogger(),
+      QUERIES_PER_SECOND,
+      MAX_QUEUED_QUERIES
+    )
+  })
+
+  test('basic message publishing passes through', async () => {
+    const message = {
+      typ: MsgType.QUERY as MsgType.QUERY,
+      id: random.randomString(32),
+      stream: FAKE_STREAM_ID,
+    }
+    const subscription = pubsub.next(message)
+    const messagePublished = new Promise<void>((resolve) => {
+      subscription.add(() => {
+        // Can be replaced with delay, but this is faster.
+        expect(ipfs.pubsub.publish).toBeCalledWith(TOPIC, serialize(message))
+        resolve()
+      })
+    })
+    await messagePublished
+    expect(ipfs.pubsub.publish).toBeCalledTimes(1)
+  })
+
+  test('Can send many non-query messages without issue', async () => {
+    const numMessages = MAX_QUEUED_QUERIES * 2
+    const messages = [...new Array(numMessages)].map(() => {
+      return {
+        typ: MsgType.UPDATE as MsgType.UPDATE,
+        stream: FAKE_STREAM_ID,
+        tip: FAKE_STREAM_ID.cid,
+      }
+    })
+
+    const allMessagesPublished = new Promise<void>((resolve, _) => {
+      let i = 0
+      messages.map((message) => {
+        pubsub.next(message).add(() => {
+          i++
+          if (i >= numMessages) {
+            resolve()
+          }
+        })
+      })
+    })
+    await allMessagesPublished
+    expect(ipfs.pubsub.publish).toBeCalledTimes(numMessages)
+  })
+
+  test('query messages are rate limited', async () => {
+    const mockNow = jest.spyOn(pubsub._clock, 'now')
+    const numMessages = QUERIES_PER_SECOND * 2
+    const messages = [...new Array(numMessages).keys()].map((i) => {
+      return {
+        typ: MsgType.QUERY as MsgType.QUERY,
+        id: i,
+        stream: FAKE_STREAM_ID,
+      }
+    })
+
+    // Fix time so it doesn't progress
+    const startTime = new Date()
+    mockNow.mockReturnValueOnce(startTime)
+    mockNow.mockReturnValueOnce(new Date(startTime.getTime() + 100))
+    mockNow.mockReturnValueOnce(new Date(startTime.getTime() + 200))
+    mockNow.mockReturnValueOnce(new Date(startTime.getTime() + 300))
+    mockNow.mockReturnValueOnce(new Date(startTime.getTime() + 400))
+    mockNow.mockReturnValue(new Date(startTime.getTime() + 500))
+
+    let maxQueriesPerSecondPublished
+    let maxQueriesPerSecondPlusOnePublished
+    const allQueriesPublished = new Promise<void>((resolveAllQueriesPublished, _) => {
```

Comment:
I really hate having to do these nested blocks just to use multiple promises at once.  Is there any nicer way to do this @ukstv?